internal/batches/executor/executor.go

```diff
@@ -69,83 +68,72 @@ type NewExecutorOpts struct {
 	IsRemote         bool
 	GlobalEnv        []string
 	ForceRoot        bool
+	FailFast         bool
 
 	BinaryDiffs bool
+	Context     context.Context
 }
 
 type executor struct {
 	opts NewExecutorOpts
 
-	par           *parallel.Run
+	workPool      *pool.ResultContextPool[*taskResult]
 	doneEnqueuing chan struct{}
-
-	results   []taskResult
-	resultsMu sync.Mutex
 }
 
 func NewExecutor(opts NewExecutorOpts) *executor {
+	p := pool.NewWithResults[*taskResult]().WithMaxGoroutines(opts.Parallelism).WithContext(opts.Context)
+	if opts.FailFast {
+		p = p.WithCancelOnError()
+	}
 	return &executor{
 		opts: opts,
 
 		doneEnqueuing: make(chan struct{}),
-		par:           parallel.NewRun(opts.Parallelism),
+		workPool:      p,
 	}
 }
 
 // Start starts the execution of the given Tasks in goroutines, calling the
 // given taskStatusHandler to update the progress of the tasks.
-func (x *executor) Start(ctx context.Context, tasks []*Task, ui TaskExecutionUI) {
+func (x *executor) Start(tasks []*Task, ui TaskExecutionUI) {
 	defer func() { close(x.doneEnqueuing) }()
 
 	for _, task := range tasks {
 		select {
-		case <-ctx.Done():
+		case <-x.opts.Context.Done():
 			return
 		default:
 		}
 
-		x.par.Acquire()
-
-		go func(task *Task, ui TaskExecutionUI) {
-			defer x.par.Release()
-
-			select {
-			case <-ctx.Done():
-				return
-			default:
-				err := x.do(ctx, task, ui)
-				if err != nil {
-					x.par.Error(err)
-				}
-			}
-		}(task, ui)
+		t := task
```

Comment:
Why create a new variable here?

Reply:
I followed the example from the sourcegraph/conc readme, and slightly recalled something about go syntax and requiring variables to be in the right scope. Not too confident if the var duplication in this example is because of the stream.

```
func process(stream chan int) {
    p := pool.New().WithMaxGoroutines(10)
    for elem := range stream {
        elem := elem
        p.Go(func() {
            handle(elem)
        })
    }
    p.Wait()
}
```

Reply:
Oh, I should fix that! That's no longer necessary after [go 1.22](https://go.dev/blog/loopvar-preview)

Reply:
Nice! Thank you :)